### ISSUE TITLE
curl_multi_exec() expects parameter 1 to be resource, int given in vendor/symfony/http-client/CurlHttpClient.php on line 323

### ISSUE THREAD

**Issue by Mrlaminat** [symfony/symfony#38129]
**Symfony version(s) affected**: 5.1.5

**Description**  
Problem with HTTP client and cURL. Throwing error:
`Fatal error: Uncaught ErrorException: curl_multi_exec() expects parameter 1 to be resource, int given in /var/www/html/vendor/symfony/http-client/CurlHttpClient.php on line 323`
while using Laravel(^7.*) internal queue worker or PHP unit tests.

**Stacktrace**
```
Call Stack:
    0.0003     401024   1. {main}() /var/www/html/vendor/phpunit/phpunit/phpunit:0
    0.0226    3408288   2. PHPUnit\TextUI\Command::main() /var/www/html/vendor/phpunit/phpunit/phpunit:61
    0.0226    3408416   3. PHPUnit\TextUI\Command->run() /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:163
    0.0475    5796624   4. PHPUnit\TextUI\TestRunner->doRun() /var/www/html/vendor/phpunit/phpunit/src/TextUI/Command.php:204
   49.4804   42338384   5. Symfony\Component\HttpClient\HttplugClient->__destruct() /var/www/html/vendor/symfony/http-client/HttplugClient.php:0
   49.4804   42338384   6. Symfony\Component\HttpClient\HttplugClient->wait() /var/www/html/vendor/symfony/http-client/HttplugClient.php:223
   49.4804   42338384   7. Symfony\Component\HttpClient\Internal\HttplugWaitLoop->wait() /var/www/html/vendor/symfony/http-client/HttplugClient.php:143
   49.4804   42338384   8. Symfony\Component\HttpClient\CurlHttpClient->stream() /var/www/html/vendor/symfony/http-client/Internal/HttplugWaitLoop.php:61
   49.4804   42338416   9. curl_multi_exec() /var/www/html/vendor/symfony/http-client/CurlHttpClient.php:323
```

From custom Service
```
/var/www/html/vendor/symfony/http-client/CurlHttpClient.php:323
/var/www/html/vendor/symfony/http-client/Internal/HttplugWaitLoop.php:61
/var/www/html/vendor/symfony/http-client/HttplugClient.php:143
/var/www/html/vendor/symfony/http-client/HttplugClient.php:223
/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php:213
/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php:192
/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/SyncQueue.php:118
/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/SyncQueue.php:47
/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Queue.php:46
/var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php:175
/var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php:158
/var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php:73
/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Bus/PendingDispatch.php:134
/var/www/html/app/Services/AttributeService.php:128
```

**Comment by nicolas-grekas** [comment]
Looking at the code, I fail to see how this property can hold an int.
Could you please provide us with a reproducer?

**Comment by Mrlaminat** [comment]
> Looking at the code, I fail to see how this property can hold an int.
> Could you please provide us with a reproducer?

I would like to provide a reproducer, but not sure that I can do it. I faced this issue with using [akeneo PHP clien](https://github.com/akeneo/api-php-client) but I think it could happen in different cases. Also in my case, I used a docker environment. What is weird for me that I tried to debug it and somehow first response to the same API works well, but the second one returns with error from cUrl

About reproducer, I could try to prepare some command samples inside the doc container for you. I will come back to it in the next few days.

**Comment by xabbuh** [comment]
@Mrlaminat Did you manage to create a reproducer? Or could you try to debug yourself how the property can end up with holding an integer?

**Comment by Mrlaminat** [comment]
The problem that I created the same container, with the same libraries versions, but couldn't reproduce it. Right now trying to debug it with Xdebug. Please don't close this branch a few days more.

**Comment by xabbuh** [comment]
Let's close this for now. We can reopen anyway once we are able to reproduce.

**Comment by nicolas-grekas** [comment]
Fixed in a580688ee5

**Comment by nicolas-grekas** [comment]
Note that I think this is a bug in PHP: curl resources are turned into integers at shutdown time.
Maybe PHP 7.4 fixed that since curl resources are now objects.
Please report back if you experience an issue on 7.4 (please upgrade if using < 7.4)

**Comment by drupol** [comment]
Hello,

I'm using PHP 7.4.12 and I think I'm having this issue, find the log [here](https://gist.github.com/drupol/6361b592628aedbfeae9c02c578f4287).

You can reproduce the issue with this [script](https://gist.github.com/drupol/89ea578c379bad7775579a13c17b5675).
The script requires [amphp/parallel-functions](https://github.com/amphp/parallel-functions).

I don't know if the issue is related to the fact that I'm using an asynchronous mechanism for requests.

Basically all the requests are sent at the same time (less or more).

**Comment by stof** [comment]
@drupol see the note about PHP resources at the beginning of the amphp/parallel-functions doc: https://amphp.org/parallel-functions/
Referencing the HttpClient in your parallel callback won't work, because the HttpClient is not meant to be serializable.
